packages/flutter/lib/src/services/text_input.dart

```diff
@@ -2513,23 +2513,46 @@ class SystemContextMenuController with SystemContextMenuClient {
   /// Null if [show] has not been called.
   Rect? _lastTargetRect;
 
+  /// The [IOSSystemContextMenuItemData]s that were last given to [show].
+  ///
+  /// Null if [show] has not been called.
+  List<IOSSystemContextMenuItemData>? _lastItems;
+
   /// True when the instance most recently [show]n has been hidden by the
   /// system.
   bool _hiddenBySystem = false;
 
-  bool get _isVisible => this == _lastShown && !_hiddenBySystem;
+  /// Indicates whether the system context menu managed by this controller is
+  /// currently being displayed to the user.
+  bool get isVisible => this == _lastShown && !_hiddenBySystem;
 
   /// After calling [dispose], this instance can no longer be used.
   bool _isDisposed = false;
 
+  final Map<int, VoidCallback> _buttonCallbacks = <int, VoidCallback>{};
+
+  /// Returns true only if the given list contains at least one pair of
+  /// identical IOSSystemContextMenuItemDatas.
+  static bool _containsDuplicates(List<IOSSystemContextMenuItemData> items) {
+    final Set<IOSSystemContextMenuItemData> uniqueItems = <IOSSystemContextMenuItemData>{};
+
+    for (final IOSSystemContextMenuItemData item in items) {
+      if (uniqueItems.contains(item)) {
+        return true;
+      }
+      uniqueItems.add(item);
+    }
+
+    return false;
+  }
+
   // Begin SystemContextMenuClient.
 
   @override
   void handleSystemHide() {
     assert(!_isDisposed);
-    // If this instance wasn't being shown, then it wasn't the instance that was
-    // hidden.
-    if (!_isVisible) {
+    assert(isVisible);
+    if (_isDisposed || !isVisible) {
```

Comment:
is it a common practice to do both assert and the if check?

Reply:
Maybe it's overkill but my reasoning is:

 * In dev mode: It should always be visible here, so we want to inform the developer if somehow it isn't.
 * In production: We want to bail here if it's not visible and not try to re-hide it.

Reply:
Is there a situation where this might hide the error in production and make it more difficult to track down?

Reply:
It's avoiding any error in production, but similar to the way that asserts are ignored in production. Is there another pattern that we would use in this situation here?